File: code/programs/rust/unix-tools/src/tr_tool.rs
//! # tr — Translate or Delete Characters
//!
//! This module implements the business logic for the `tr` command.
//! tr operates on individual characters, translating, deleting, or
//! squeezing them according to two character set specifications.
//!
//! ## Operations
//!
//! ```text
//!     tr 'a-z' 'A-Z'         Translate lowercase to uppercase
//!     tr -d 'aeiou'           Delete all vowels
//!     tr -s 'a-z'             Squeeze repeated lowercase chars
//!     tr -c 'a-z' '_'         Replace non-lowercase with '_'
//! ```
//!
//! ## Character Sets and Ranges
//!
//! tr supports character ranges like `a-z` (all lowercase letters)
//! and `0-9` (all digits). Characters in SET1 are mapped positionally
//! to characters in SET2.

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Expand a character set specification into individual characters.
///
/// Handles range notation like "a-z" by expanding it to all characters
/// in that range. For example: "a-d" becomes "abcd".
pub fn expand_set(set: &str) -> Vec<char> {
    let chars: Vec<char> = set.chars().collect();
    let mut result: Vec<char> = Vec::new();
    let mut i = 0;

    while i < chars.len() {
        // Check for range notation: "a-z"
        if i + 2 < chars.len() && chars[i + 1] == '-' {
            let start = chars[i];
            let end = chars[i + 2];
            if start <= end {
                for c in start..=end {
                    result.push(c);
                }
            } else {
                // Reverse range — treat as literals.
                result.push(chars[i]);
                result.push(chars[i + 1]);
                result.push(chars[i + 2]);
            }
            i += 3;
        } else if chars[i] == '\\' && i + 1 < chars.len() {
            // Handle escape sequences.
            match chars[i + 1] {
                'n' => result.push('\n'),
                't' => result.push('\t'),
                'r' => result.push('\r'),
                '\\' => result.push('\\'),
                other => result.push(other),
            }
            i += 2;
        } else {
            result.push(chars[i]);
            i += 1;
        }
    }

    result
}

/// Translate characters: replace each char in SET1 with the corresponding
/// char in SET2.
///
/// If SET2 is shorter than SET1, the last character of SET2 is used for
/// the remaining mappings.
pub fn translate(content: &str, set1: &str, set2: &str, complement: bool) -> String {
    let exp1 = expand_set(set1);
    let exp2 = expand_set(set2);

    content
        .chars()
        .map(|ch| {
            let in_set = exp1.contains(&ch);
            let should_translate = if complement { !in_set } else { in_set };

            if should_translate && !exp2.is_empty() {
                if complement {
                    exp2[0]
                } else if let Some(idx) = exp1.iter().position(|&c| c == ch) {
                    if idx < exp2.len() {
                        exp2[idx]
                    } else {
                        *exp2.last().unwrap()
                    }
                } else {
                    ch
                }
            } else {
                ch
            }
        })
        .collect()
}

/// Delete characters in SET1 from the content.
pub fn delete_chars(content: &str, set1: &str, complement: bool) -> String {
    let expanded = expand_set(set1);

    content
        .chars()
        .filter(|ch| {
            let in_set = expanded.contains(ch);
            if complement { in_set } else { !in_set }
        })
        .collect()
}

/// Squeeze adjacent repeated characters that appear in the set.
///
/// Replaces runs of the same character with a single occurrence,
/// but only for characters that are in the specified set.
pub fn squeeze_repeats(content: &str, set: &str) -> String {
    let expanded = expand_set(set);
    let mut result = String::new();
    let mut last_char: Option<char> = None;

    for ch in content.chars() {
        if let Some(prev) = last_char {
            if ch == prev && expanded.contains(&ch) {
                continue; // Skip repeated character
            }
        }
        result.push(ch);
        last_char = Some(ch);
    }

    result
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expand_range() {
        let result = expand_set("a-d");
        assert_eq!(result, vec!['a', 'b', 'c', 'd']);
    }

    #[test]
    fn expand_literal() {
        let result = expand_set("abc");
        assert_eq!(result, vec!['a', 'b', 'c']);
    }

    #[test]
    fn expand_mixed() {
        let result = expand_set("a-c0-2");
        assert_eq!(result, vec!['a', 'b', 'c', '0', '1', '2']);
    }

    #[test]
    fn translate_simple() {
        assert_eq!(translate("hello", "helo", "HELO", false), "HELLO");
    }

    #[test]
    fn translate_range() {
        let result = translate("abc", "a-c", "A-C", false);
        assert_eq!(result, "ABC");
    }

    #[test]
    fn delete_vowels() {
        assert_eq!(delete_chars("hello world", "aeiou", false), "hll wrld");
    }

    #[test]
    fn delete_complement() {
        // Delete everything that is NOT a vowel.
        assert_eq!(delete_chars("hello", "helo", true), "");
    }

    #[test]
    fn squeeze_simple() {
        assert_eq!(squeeze_repeats("aabbcc", "a-c"), "abc");
    }

    #[test]
    fn squeeze_partial() {
        // Only squeeze 'a', not 'b'.
        assert_eq!(squeeze_repeats("aabb", "a"), "abb");
    }

    #[test]
    fn empty_input() {
        assert_eq!(translate("", "a", "b", false), "");
        assert_eq!(delete_chars("", "a", false), "");
        assert_eq!(squeeze_repeats("", "a"), "");
    }
}

File: code/programs/go/unix-tools/tr_tool.go
// =========================================================================
// tr — Translate or Delete Characters
// =========================================================================
//
// The `tr` utility translates, squeezes, or deletes characters from
// standard input. It operates on individual characters, not words or lines.
//
// # Basic usage
//
//   echo "hello" | tr 'a-z' 'A-Z'     Convert lowercase to uppercase
//   echo "hello" | tr -d 'l'          Delete all 'l' characters => "heo"
//   echo "aabbcc" | tr -s 'a-z'       Squeeze repeated chars => "abc"
//   echo "hello" | tr -c 'a-z' '_'    Replace non-lowercase with '_'
//
// # Character sets
//
// tr takes two character set arguments:
//   SET1: characters to translate from (or delete)
//   SET2: characters to translate to
//
// Characters are mapped positionally: SET1[0] -> SET2[0], SET1[1] -> SET2[1], etc.
//
// # Special notations
//
// tr supports ranges like 'a-z' and some POSIX classes. For simplicity,
// this implementation handles:
//   - Literal characters: "abc"
//   - Ranges: "a-z", "A-Z", "0-9"
//
// # Architecture
//
//   tr.json (spec)              tr_tool.go (this file)
//   ┌──────────────────┐      ┌────────────────────────────────┐
//   │ flags: -c,-d,-s  │      │ read stdin char by char         │
//   │ -t               │─────>│ translate/delete/squeeze        │
//   │ args: SET1,SET2  │      │ write to stdout                │
//   └──────────────────┘      └────────────────────────────────┘

package main

import (
	"fmt"
	"io"
	"os"
	"strings"

	clibuilder "github.com/adhithyan15/coding-adventures/code/packages/go/cli-builder"
)

// =========================================================================
// expandSet — expand a character set specification into individual chars
// =========================================================================
//
// This function expands range notations like "a-z" into the full set of
// characters. For example:
//   "a-z"    => "abcdefghijklmnopqrstuvwxyz"
//   "a-d0-3" => "abcd0123"
//   "abc"    => "abc"
//
// The expansion algorithm:
//   1. Walk through the string character by character
//   2. When we see a '-' between two characters, expand the range
//   3. Otherwise, add the literal character

func expandSet(set string) string {
	runes := []rune(set)
	var result []rune

	i := 0
	for i < len(runes) {
		// Check for range notation: "a-z"
		// A range requires at least 3 characters: start, '-', end
		if i+2 < len(runes) && runes[i+1] == '-' {
			start := runes[i]
			end := runes[i+2]

			// Expand the range from start to end (inclusive).
			if start <= end {
				for c := start; c <= end; c++ {
					result = append(result, c)
				}
			} else {
				// Reverse range (e.g., "z-a") — treat as literals.
				result = append(result, runes[i], runes[i+1], runes[i+2])
			}
			i += 3
		} else {
			// Handle backslash escapes for common special characters.
			if runes[i] == '\\' && i+1 < len(runes) {
				switch runes[i+1] {
				case 'n':
					result = append(result, '\n')
					i += 2
					continue
				case 't':
					result = append(result, '\t')
					i += 2
					continue
				case 'r':
					result = append(result, '\r')
					i += 2
					continue
				case '\\':
					result = append(result, '\\')
					i += 2
					continue
				}
			}
			result = append(result, runes[i])
			i++
		}
	}

	return string(result)
}

// =========================================================================
// translateContent — translate characters in content using SET1 and SET2
// =========================================================================
//
// This function maps each character in the input: if it appears in SET1,
// replace it with the corresponding character in SET2. If SET2 is shorter
// than SET1, the last character of SET2 is used for the remaining mappings.

func translateContent(content string, set1, set2 string, complement bool) string {
	expandedSet1 := expandSet(set1)
	expandedSet2 := expandSet(set2)

	var result strings.Builder

	for _, ch := range content {
		idx := strings.IndexRune(expandedSet1, ch)
		inSet := idx >= 0

		// With -c (complement), we translate characters NOT in SET1.
		if complement {
			inSet = !inSet
		}

		if inSet {
			if len(expandedSet2) > 0 {
				set2Runes := []rune(expandedSet2)
				if complement {
					// For complement mode, always use the first char of SET2.
					result.WriteRune(set2Runes[0])
				} else if idx < len(set2Runes) {
					result.WriteRune(set2Runes[idx])
				} else {
					// SET2 is shorter — use the last character.
					result.WriteRune(set2Runes[len(set2Runes)-1])
				}
			}
		} else {
			result.WriteRune(ch)
		}
	}

	return result.String()
}

// =========================================================================
// deleteChars — delete characters in SET1 from content
// =========================================================================

func deleteChars(content, set1 string, complement bool) string {
	expanded := expandSet(set1)
	var result strings.Builder

	for _, ch := range content {
		inSet := strings.ContainsRune(expanded, ch)
		if complement {
			inSet = !inSet
		}
		if !inSet {
			result.WriteRune(ch)
		}
	}

	return result.String()
}

// =========================================================================
// squeezeRepeats — replace runs of repeated characters with single occurrences
// =========================================================================

func squeezeRepeats(content, set string) string {
	expanded := expandSet(set)
	var result strings.Builder
	var lastChar rune
	first := true

	for _, ch := range content {
		if !first && ch == lastChar && strings.ContainsRune(expanded, ch) {
			// Skip repeated characters that are in the squeeze set.
			continue
		}
		result.WriteRune(ch)
		lastChar = ch
		first = false
	}

	return result.String()
}

// =========================================================================
// runTr — the testable core of the tr tool
// =========================================================================

func runTr(specPath string, argv []string, stdout io.Writer, stderr io.Writer) int {
	return runTrWithStdin(specPath, argv, stdout, stderr, os.Stdin)
}

// runTrWithStdin is the inner implementation that accepts a custom stdin.

func runTrWithStdin(specPath string, argv []string, stdout io.Writer, stderr io.Writer, stdin io.Reader) int {
	// Step 1: Create the parser from the spec.
	parser, err := clibuilder.NewParser(specPath, argv)
	if err != nil {
		fmt.Fprintf(stderr, "tr: %s\n", err)
		return 1
	}

	// Step 2: Parse the arguments.
	result, err := parser.Parse()
	if err != nil {
		fmt.Fprintf(stderr, "%s\n", err)
		return 1
	}

	// Step 3: Handle the result.
	switch r := result.(type) {

	case *clibuilder.HelpResult:
		fmt.Fprintln(stdout, r.Text)
		return 0

	case *clibuilder.VersionResult:
		fmt.Fprintln(stdout, r.Version)
		return 0

	case *clibuilder.ParseResult:
		// Extract flags.
		complement := getBool(r.Flags, "complement")
		deleteMode := getBool(r.Flags, "delete")
		squeeze := getBool(r.Flags, "squeeze_repeats")

		// Extract SET1 and SET2.
		set1, _ := r.Arguments["set1"].(string)
		set2, _ := r.Arguments["set2"].(string)

		// Read all input from stdin.
		inputBytes, err := io.ReadAll(stdin)
		if err != nil {
			fmt.Fprintf(stderr, "tr: error reading input: %s\n", err)
			return 1
		}
		content := string(inputBytes)

		// Apply the requested transformation.
		var output string

		if deleteMode {
			// -d: delete characters in SET1.
			output = deleteChars(content, set1, complement)
			// If -s is also set, squeeze SET2 characters.
			if squeeze && set2 != "" {
				output = squeezeRepeats(output, set2)
			}
		} else if squeeze && set2 == "" {
			// -s without SET2: squeeze characters in SET1.
			output = squeezeRepeats(content, set1)
		} else {
			// Translate mode: replace SET1 chars with SET2 chars.
			output = translateContent(content, set1, set2, complement)
			// If -s is also set, squeeze SET2 characters.
			if squeeze {
				output = squeezeRepeats(output, set2)
			}
		}

		fmt.Fprint(stdout, output)
		return 0

	default:
		fmt.Fprintf(stderr, "tr: unexpected result type: %T\n", result)
		return 1
	}
}
